damnit/ctxsupport/damnit_ctx.py

```diff
@@ -128,6 +157,301 @@ def evaluate(self, run_data, kwargs):
 
         return cell
 
+    def copy(self):
+        return copy(self)
+
+
+def is_group(obj) -> bool:
+    """Return True if obj is a Group class or an instance of a Group class."""
+    cls = obj if isinstance(obj, type) else type(obj)
+    return hasattr(cls, GROUP_MARKER_ATTR)
+
+
+def is_group_instance(obj) -> bool:
+    """Return True if obj is an instance of a Group class.
+    """
+    return hasattr(type(obj), GROUP_MARKER_ATTR)
+
+
+class Group:
+    """A class decorator that transforms a class into a Group.
+
+    Transforms a class into a Group 'dataclass'. Allow for creating reusable,
+    configurable, and nestable sets of related Variables.
+
+    - `@Group` for default behavior.
+    - `@Group(title='...', ...)` to provide parameters.
+
+    Args:
+        title (str, optional): The title of the group.
+        tags (list[str], optional): A list of tags to associate with the group.
+        sep (str, optional): Separator to use between group title and variable title.
+    """
+    # _REGISTRY: WeakValueDictionary[str, 'Group'] = WeakValueDictionary()
+
+    def __new__(cls, decorated_cls: type = None, **kwargs) -> type:
+        def wrapper(wrapped_cls: type) -> type:
+            return _new_group(wrapped_cls, kwargs)
+
+        return wrapper if decorated_cls is None else wrapper(decorated_cls)
+
+
+@dataclass
+class _GroupBase:
+    name: str | None = field(default=None, kw_only=True)
+    title: str | None = field(default=None, kw_only=True)
+    tags: tuple[str] | None = field(default=None, kw_only=True)
+    sep: str = field(default='/', kw_only=True)
+    # set of vars that were removed due to missing dependencies
+    _removed_vars: set[str] = field(default_factory=set, init=False)
+
+    def __post_init__(self):
+        """Post-initialization to ensure tags are always a tuple."""
+        if isinstance(self.tags, str):
+            self.tags = (self.tags,)
+
+        # register group instance
+        if self.name is None:
+            self.name = self.__class__.__name__
+
+        if self.title is None:
+            self.title = self.name
+
+        # Modify variables with group context
+        self._morph_vars()
+
+    def __hash__(self):
+        return hash(self.name)
+
+    def __getattribute__(self, item):
+        removed_vars = super().__getattribute__('_removed_vars')
+        if item in removed_vars:
+            raise AttributeError(
+                f"Variable {item!r} was removed from group "
+                f"{super().__getattribute__('name')!r} due to missing dependencies"
+            )
```

Comment:
This mechanism feels like quite a bit of extra complexity for something that I'd hope is possible with either group-based subclassing or the existing mechanism to skip variables with missing dependencies. Can we work out if we really need yet another piece for this?